### ISSUE TITLE
Remove keyword duplication such as same keywords with different capitalization

### ISSUE THREAD

**Issue by hyi** [hydroshare/hydroshare#3563]
An example with duplicate keywords are in this resource: https://www.hydroshare.org/resource/c71e41fe816548889ee039e870098eb4/

**Comment by hyi** [comment]
Have investigated this issue a bit and found we are already doing insensitive comparison when adding keywords as shown here: https://github.com/hydroshare/hydroshare/blob/develop/hs_core/models.py#L1592.

I believe the problem with that duplicate keyword issue as evidenced in this resource: https://www.hydroshare.org/resource/c71e41fe816548889ee039e870098eb4/ should be caused by the multiple or double spaces between words in a keyword rather than the capitalization. Have updated the title of this issue to reflect this. 

To resolve this without too much performance compromise, I think we will need to do database raw queries to use ```replace``` function to remove all spaces within a keyword phrase and compare against the new keyword phrase with spaces removed as well, or do regular expression query to achieve the same comparison with all spaces within removed. This will definitely introduce a bit of complexity, so want to get input to see whether this issue is important enough to justify the complexity  that will be introduced into our code base. @dtarb @martinseul @horsburgh @sblack-usu @pkdash

**Comment by sblack-usu** [comment]
The specific example you cited isn't due to multiple spaces between words, it's a trailling space at the beginning of the keyword
![image](https://user-images.githubusercontent.com/35820390/64046124-477cc380-cb28-11e9-8e6b-d0abf74caad0.png)

**Comment by hyi** [comment]
@sblack-usu Thanks for the info. That definitely will need to be fixed - the keyword value needs to be stripped to remove the leading and trailing spaces before being added, which is not done currently. I can easily create a PR to fix this, but do we want to also fix the multiple spaces between words issue? I am thinking that will only occur very rarely potentially by mistake, and users can remove extra duplicate keywords as needed, so the effort to spend to fix multiple space between keywords issue does not sound justified to me. If you agree, I will just create a PR to fix this leading and trailing space issue without spending time on the potential issue with multiple spaces between words.

**Comment by sblack-usu** [comment]
I think if this needs to be fixed, the keywords should be normalized when saved to the db.  So something like a regular expression to replace 2 or more spaces with a single space and then call strip()?

If you do that, then you'll have to also normalize all of the existing keywords too.  You won't be able to do it just on the db because I think solr and the bag files will need to be updated as well.  So I think a management command would be best.

**Comment by hyi** [comment]
@sblack-usu Good points, and agreed. Looks like a decent amount of work would be involved to resolve this completely. I'll wait for others to chime in to see whether this is worth the effort @dtarb @martinseul

**Comment by dtarb** [comment]
I would characterize this as a cosmetic issue with very low priority - especially given the complexity you have identified.  If a user does not like duplicate keywords, they can clean them themselves easily.  There many other more critical issues and bugs to fix first.

**Comment by cuahsimarko** [comment]
@hyi have you reviewed the frontend code that pre-processes the keywords? I believe some of this is handled there.

**Comment by hyi** [comment]
@cuahsimarko Good point. I'll take a look at the frontend code.

**Comment by sblack-usu** [comment]
If it is handled in the frontend, it should be done in the backend so adding keywords via the rest api behaves in the same way.

**Comment by hyi** [comment]
@sblack-usu Agreed, which is why I did not look at the frontend code at all in the first place. Will look at the frontend code for keywords preprocessing logic and move/consolidate into backend if any.

**Comment by cuahsimarko** [comment]
@hyi @sblack-usu we should look into this as a team. Frontend validation has always had inline user feedback. Stripping that out and moving validation to a pure backend function would result in a poorer UX

**Comment by sblack-usu** [comment]
I don't mind if the UX does checks and gives user feedback, but it should be consistent with the backend.

**Comment by sblack-usu** [comment]
I think we should also keep in mind that this issue is low priority.  Though I do think the conversations around front/back end validations that have come out of this are worth having.

**Comment by hyi** [comment]
@cuahsimarko @sblack-usu I'll look into frontend code and see if I can taken into consideration both UX and backend validation for consistency and REST API access purpose. I know we have been doing both for many cases in HydroShare. I think when UX is not much affected, it'd be good to leave validation to backend so that we have single point of truth for validation. However, if UX were affected by this, some quick and fast validation on the frontend would make sense to me.

**Comment by cuahsimarko** [comment]
> I think we should also keep in mind that this issue is low priority. Though I do think the conversations around front/back end validations that have come out of this are worth having.

Totally agree. I'm happy to move this conversation out of this Issue so as not to pollut the conversation, but overall inline/immediate userfeedback for more complex inputs, autocomplete and other contemporary UI elements have been on my mind. I'd like to start by enumerating areas for "inline feedback/validation text" and "toasts that display a message for n seconds"... I'll take the action item to briefly document and schedule a kickoff meeting at some point with low-moderate priority.

**Comment by cuahsimarko** [comment]
Sure frontend and backend validation can overlap and should never conflict in logic.
